cpp/daal/src/algorithms/dtrees/gbt/classification/gbt_classification_predict_dense_default_batch_impl.i

```diff
@@ -178,8 +178,103 @@ public:
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    using Dispetcher_t = gbt::prediction::internal::PredictDispetcher<hasUnorderedFeatures, hasAnyMissing>;
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    void predictByTrees(algorithmFPType * val, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x,
+                        const Dispetcher_t<hasUnorderedFeatures, hasAnyMissing>& dispetcher);
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    void predictByTreesVector(algorithmFPType * val, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x,
+                              const Dispetcher_t<hasUnorderedFeatures, hasAnyMissing>& dispetcher);
+
+    template <bool val>
+    struct BooleanConstant {
+        typedef BooleanConstant<val> type;
+    };
+
+    inline void updateResult(algorithmFPType * res, algorithmFPType * val, size_t iRow, size_t i, size_t nClasses, BooleanConstant<true> dispetcher) {
+        res[iRow + i] = algorithmFPType(getMaxClass(val + i * nClasses, nClasses));
+    }
+
+    inline void updateResult(algorithmFPType * res, algorithmFPType * val, size_t iRow, size_t i, size_t nClasses, BooleanConstant<false> dispetcher) {}
+
+    inline algorithmFPType* updateBuffer(algorithmFPType * val, size_t buf_shift, size_t buf_size, BooleanConstant<true> dispetcher) {
+        services::internal::service_memset_seq<algorithmFPType, cpu>(val, algorithmFPType(0), buf_size);
+        return val;
+    }
+
+    inline algorithmFPType* updateBuffer(algorithmFPType * val, size_t buf_shift, size_t buf_size, BooleanConstant<false> dispetcher) {
+        return val + buf_shift;
+    }
+
+    inline bool checkForMissing(const algorithmFPType* x, size_t nTrees, size_t nRows, size_t nColumns) {
+         size_t nLvlTotal = 0;
+        for (size_t iTree = 0; iTree < nTrees; ++iTree) {
+           nLvlTotal += this->_aTree[iTree]->getMaxLvl();
+        }
+        if (nLvlTotal <= nColumns) {
+            // Checking is compicated. Better to do it during inferense.
+            return true;
+        } else {
+            for (size_t idx = 0; idx < nRows * nColumns; ++idx) {
+                if (isnan(x[idx])) return true;
+            }
+        }
+        return false;
+    }
+
+    template <bool hasUnorderedFeatures, bool hasAnyMissing, bool isResValidPtr, bool reuseBuffer>
+    inline void predict(size_t nTrees, size_t nClasses, size_t nRows, size_t nColumns,
+                        const algorithmFPType * x, algorithmFPType * valL, algorithmFPType * res) {
+        Dispetcher_t<hasUnorderedFeatures, hasAnyMissing> dispetcher;
+        size_t iRow = 0;
+        for (; iRow + VECTOR_BLOCK_SIZE <= nRows; iRow += VECTOR_BLOCK_SIZE)
+        {
+            algorithmFPType* val = updateBuffer(valL, iRow * nClasses, nClasses * VECTOR_BLOCK_SIZE, BooleanConstant<reuseBuffer>());
```

Comment:
Isn't it simpler to just use a ternary expression here?

```c++
val = reuseBuffer ? valL + iRow * nClasses : services::internal::service_memset_seq<algorithmFPType, cpu>(valL, algorithmFPType(0), iRow * nClasses );
```

Reply:
In the best of my understanding, in your realization we can't be sure in optimization of `if `statement by compiler. Am I wrong?